src/vmm/src/pipe.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright 2020 Sartura All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+use std::fs::{File, OpenOptions};
+use std::io::{Read, Write};
+use std::os::unix::io::{AsRawFd, RawFd};
+use std::path::PathBuf;
+
+/// A wrapper structure around that provides basic read functionality backed by
+/// a named pipe implementation.
+pub(crate) struct PipeReaderWrapper(PathBuf, File);
+
+impl PipeReaderWrapper {
+    pub fn new(buf: PathBuf) -> PipeReaderWrapper {
+        match utils::mkfifo(&buf, libc::S_IRWXO) {
+            Ok(_) => (),
+            Err(err) => panic!("Error creating fifo: {}", err),
+        }
+        // let pipe = match File::open(&buf) {
+        let mut options = OpenOptions::new();
+        let pipe = match options.read(true).write(true).create(true).open(&buf) {
+            Err(err) => panic!("couldn't open pipe: {}", err),
+            Ok(output) => output,
+        };
+        PipeReaderWrapper(buf, pipe)
+    }
+}
+
+impl Read for PipeReaderWrapper {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        self.1.read(buf)
+    }
+}
+
+impl AsRawFd for PipeReaderWrapper {
+    fn as_raw_fd(&self) -> RawFd {
+        self.1.as_raw_fd()
+    }
+}
+
+/// A wrapper structure around that provides basic write functionality backed by
+/// a named pipe implementation.
+pub(crate) struct PipeWriterWrapper(PathBuf);
+
+impl PipeWriterWrapper {
+    pub fn new(buf: PathBuf) -> PipeWriterWrapper {
+        match utils::mkfifo(&buf, libc::S_IRWXO) {
+            Ok(_) => (),
+            Err(err) => panic!("Error creating fifo: {}", err),
+        }
+        PipeWriterWrapper(buf)
+    }
+}
+
+impl Write for PipeWriterWrapper {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        let mut pipe = match File::create(&self.0) {
```

Comment:
I don't understand this part. Why do we need to call `File::create` here? Isn't the file already created when calling new? Shouldn't we be able to just call `open` on the `PathBuf`?

Reply:
```File::create``` opens a file in write-only mode, whereas ```File::open``` opens a file in read-only mode. Since this is the writer this seems fine. I agree that it's weirdly named, but that's the way it's [documented](https://doc.rust-lang.org/std/fs/struct.File.html#method.create).